run.sh

```diff
@@ -1,116 +1,345 @@
+    ## parse command arguments
 
-else
-    # otherwise, a simple proxy to pytest
-    pytest -p no:warnings $@
+    # parse flags
+    while [[ "$#" -gt 0 ]]; do
+        case "${1}" in
+            -h|--help)
+                help
+                exit
+                ;;
+            -d|--docker)
+                docker=1
+                ;;
+            -G|--scenario-group)
+                # TODO: get group
+                # upcase via ${2^^} is unsupported on bash 3.x
+                mapfile -t group <<< "$(lookup_scenario_group "$(echo "${2}" | upcase)")"
+                scenarios+=("${group[@]}")
+                shift
+                ;;
+            -S|--scenario)
+                # upcase via ${2^^} is unsupported on bash 3.x
+                scenarios+=("$(echo "${2}" | upcase)")
+                shift
+                ;;
+
+            --)
+                # ignore and stop flag processing to force remainder to be captured as is
+                shift
+                break
+                ;;
+            -*)
+                # unknown flag: be helpful
+                error "unknown flag: ${1}"
+                help
+                exit 64
+                ;;
+            *)
+                # handle positional arguments
+                # deprecated but kept for backwards compatibility
+                if [[ "${1}" =~ [A-Z0-9_]+_SCENARIOS$ ]]; then
+                    # TODO: get group
+                    scenarios+=("${1}")
+                elif [[ "${1}" =~ ^[A-Z0-9_]+$ ]]; then
+                    scenarios+=("${1}")
+                else
+                  error "invalid argument: ${1}"
+                  help
+                  exit 64
+                fi
+                ;;
+        esac
+        shift
+    done
+
+    # capture remainder of arguments to pass as-is for pytest
+    pytest_args+=("${@}")
+
+    ## prepare commands
+
+    # when no scenario is provided, use a nice default
+    if [[ "${#scenarios[@]}" -lt 1 ]]; then
+        scenarios+=('DEFAULT')
+    fi
+
+    # backward compatibility with scenarios that have been removed/renamed
+    # TODO: remove once all CIs have been updated
+    for i in "${!scenarios[@]}"; do
+        case "${scenarios["${i}"]}" in
+            APPSEC_IP_BLOCKING_MAXED|APPSEC_IP_BLOCKING)
+                scenarios+=(APPSEC_BLOCKING_FULL_DENYLIST)
+                unset "scenarios[${i}]"
+                ;;
+        esac
+    done
+
+    # TODO: remove duplicates
+
+    # TODO: upgrade the dependencies to the latest version of pulumi once the protobuf bug is fixed
+    # In the meantime remove the warning from the output
+    pytest_args+=( '-p' 'no:warnings' )
+
+    ## run tests
+
+    if [[ "${docker}" == 1 ]]; then
+        run_mode='docker'
+    else
+        run_mode='direct'
+
+        # cleanups
+        clean_pycache
+
+        # ensure environment
+        if ! is_using_nix; then
+            activate_venv
+        fi
+    fi
+
+    echo "Run plan:"
```

Comment:
Do not add anything in sdtout here, pytest must be the only process to handle the ouput.

Reply:
Got it.